File: SuitePlugins/MapPacker/src/main/java/old/MapPacker.java
//package old;
//
//import com.google.common.collect.Lists;
//import com.google.common.collect.Maps;
////import com.google.gson.Gson;
////import com.google.gson.GsonBuilder;
//
//import javafx.fxml.FXML;
//import javafx.scene.canvas.Canvas;
//import javafx.scene.control.*;
//import javafx.scene.layout.AnchorPane;
//import javafx.scene.layout.HBox;
//import javafx.scene.layout.VBox;
//
//import lombok.extern.slf4j.Slf4j;
//import misc.CustomTab;
//import store.CacheLibrary;
//import store.cache.index.OSRSIndices;
//import store.cache.index.archive.Archive;
//import store.plugin.extension.FXController;
//import store.utilities.GZIPCompressor;
//import suite.Constants;
//import suite.Main;
//import suite.controller.Selection;
//import suite.dialogue.Dialogues;
//import utility.*;
//
//import java.io.File;
//import java.nio.file.Files;
//import java.util.*;
//import java.util.stream.Collectors;
//
///**
// * @author ReverendDread & RSPSi on 11/27/2019
// * @project ValkyrCacheSuite
// */
//@Slf4j
//public class MapPacker extends FXController {
//
//	private CustomTab tab;
//	@FXML
//	private RadioButton folderRadio;
//
//	@FXML
//	private ToggleGroup loadType;
//
//	@FXML
//	private VBox folderFileBox;
//
//	@FXML
//	private TextField mapFolderText;
//
//	@FXML
//	private Button mapFolderBrowse;
//
//	@FXML
//	private RadioButton singleFileRadio;
//
//	@FXML
//	private VBox singleFileBox;
//
//	@FXML
//	private TextField lsText;
//
//	@FXML
//	private Button lsBrowse;
//
//	@FXML
//	private TextField objText;
//
//	@FXML
//	private Button objBrowse;
//
//	@FXML
//	private RadioButton regionIdRadio;
//
//	@FXML
//	private ToggleGroup regionType;
//
//	@FXML
//	private HBox regionIdBox;
//
//	@FXML
//	private TextField regionID;
//
//	@FXML
//	private RadioButton regionXYRadio;
//
//	@FXML
//	private HBox regionXYBox;
//
//	@FXML
//	private TextField regionXText;
//
//	@FXML
//	private TextField regionYText;
//
//	@FXML
//	private RadioButton worldRadio;
//
//	@FXML
//	private HBox worldXYBox;
//
//	@FXML
//	private TextField worldXText;
//
//	@FXML
//	private TextField worldYText;
//
//	@FXML
//	private TextField xteaText;
//
//	@FXML
//	private Button genXteas;
//
//	@FXML
//	private Button saveBtn;
//
//	@FXML
//	private Canvas canvas;
//
//	@FXML
//	private AnchorPane worldAnchor;
//
//	@FXML
//	private TextField regionIdField;
//
//	@FXML
//	private Button regionIdSubmit;
//
//	private WorldMapRenderer renderer;
//
//    private List<XTEA> xteaStore = Lists.newArrayList();
//
////    private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
//
////    public void saveXTeas(){
////        if(Constants.settings.xteaFile == null || Constants.settings.xteaFile.isEmpty())
////            return;
////
////        try(FileWriter fw = new FileWriter(new File(Constants.settings.xteaFile))){
////            GSON.toJson(xteaStore, fw);
////        } catch (Exception e) {
////            e.printStackTrace();
////        }
////    }
//
//	@Override
//	public void initialize(CustomTab tab, boolean refresh, int lastId) {
//        if (Constants.settings.xteaFile == null || Constants.settings.xteaFile.isEmpty()) {
//            File file = RetentionFileChooser.showOpenDialog("Please select xteas.json", Main.getSelection().stage.getScene().getWindow(), FilterMode.JSON);
//            if (file != null) {
//                Constants.settings.xteaFile = file.getAbsolutePath();
//                Constants.settings.save();
//            }
//        }
//
////        if(Constants.settings.xteaFile != null && !Constants.settings.xteaFile.isEmpty()){
////            try(FileReader fr = new FileReader(new File(Constants.settings.xteaFile))) {
////                xteaStore = GSON.fromJson(fr, new TypeToken<List<XTEA>>() { }.getType());
////            } catch (Exception ex){
////                ex.printStackTrace();
////            }
////        }
//
//		renderer = new WorldMapRenderer(canvas);
//		new Thread(renderer).start();
//
//		XTEASManager.get().setParserType(XTEASManager.XTEAParserType.POLAR);
//
//	    regionType.selectedToggleProperty().addListener((observable, oldValue, newValue) -> {
//            regionIdBox.setDisable(newValue != regionIdRadio);
//            regionXYBox.setDisable(newValue != regionXYRadio);
//            worldXYBox.setDisable(newValue != worldRadio);
//        });
//
//	    loadType.selectedToggleProperty().addListener((observable, oldValue, newValue) -> {
//            folderFileBox.setDisable(newValue != folderRadio);
//            singleFileBox.setDisable(newValue != singleFileRadio);
//        });
//		this.tab = tab;
//
//		mapFolderBrowse.setOnAction(evt -> {
//            File file = RetentionFileChooser.showOpenDialog(Main.getSelection().stage.getScene().getWindow(), FilterMode.PACK);
//            if (file != null) {
//                mapFolderText.setText(file.getAbsoluteFile().toString());
//            }
//        });
//
//		lsBrowse.setOnAction(evt -> {
//			File file = RetentionFileChooser.showOpenDialog(Main.getSelection().stage.getScene().getWindow(), FilterMode.DAT, FilterMode.GZIP);
//			if (file != null) {
//				lsText.setText(file.getAbsoluteFile().toString());
//			}
//		});
//
//		objBrowse.setOnAction(evt -> {
//			File file = RetentionFileChooser.showOpenDialog(Main.getSelection().stage.getScene().getWindow(), FilterMode.DAT, FilterMode.GZIP);
//			if (file != null) {
//				objText.setText(file.getAbsoluteFile().toString());
//			}
//		});
//
//		saveBtn.setOnAction(evt -> {
//
//			int regionX, regionY;
//			int regionId = Integer.parseInt(regionID.getText());
//			if (regionType.getSelectedToggle() == regionIdRadio) {
//				regionX = regionId >> 8;
//				regionY = regionId & 0xFF;
//			} else if (regionType.getSelectedToggle() == regionXYRadio) {
//				regionX = Integer.parseInt(regionXText.getText());
//				regionY = Integer.parseInt(regionYText.getText());
//			} else {
//				regionX = (int) Math.floor(Integer.parseInt(worldXText.getText()) / 64.0);
//				regionY = (int) Math.floor(Integer.parseInt(worldYText.getText()) / 64.0);
//			}
//
//			int[] xteas = xteaText.getText().isEmpty() ? XTEASManager.lookup(regionId) : Arrays.stream(xteaText.getText().replace(" ", "").split(",")).mapToInt(Integer::parseInt).toArray();
//
//			if (Objects.isNull(xteas)) {
//				xteas = generateXTEAS();
//				log.info("Generating xteas for region {}, {}", regionId, xteas);
//			}
//
//			if (xteas.length < 4) {
//				Dialogues.alert(Alert.AlertType.ERROR, "XTEA Format Incorrect", "Please check your xteas format", "needs to be - 0, 0, 0, 0", false);
//				return;
//			}
//
//			Map<Integer, int[]> xteaKeys = Maps.newConcurrentMap();
//
//			if (loadType.getSelectedToggle() == folderRadio) {
//				if (mapFolderText.getText().isEmpty()) {
//					Alert alert = new Alert(Alert.AlertType.ERROR, "Please select map files!", ButtonType.OK);
//					alert.show();
//					return;
//				}
//				if (!(mapFolderText.getText().endsWith(".pack"))) {
//					Alert alert = new Alert(Alert.AlertType.ERROR, "File needs to be .pack format!", ButtonType.OK);
//					alert.show();
//					return;
//				}
//				try {
//					List<MultiMapEncoder.Chunk> chunks = MultiMapEncoder.decode(Files.readAllBytes(new File(mapFolderText.getText()).toPath()));
//					int[] finalXteas = xteas;
//					chunks.stream()
//					.map(chunk -> {
//						int localRegionId = (regionX + chunk.offsetX) << 8 | (regionY + chunk.offsetY);
//						xteaStore.removeIf(xtea -> xtea.getRegion() == localRegionId);
//						xteaStore.add(new XTEA(localRegionId, finalXteas));
//						return packMaps(regionX + (chunk.offsetX), regionY + (chunk.offsetY), chunk.tileMapData, chunk.objectMapData, finalXteas);
//					})
//					.forEach(map -> map.entrySet().forEach(entry -> xteaKeys.put(entry.getKey(), entry.getValue())));
//				} catch (Exception ex) {
//					ex.printStackTrace();
//				}
//			} else {
//				if (lsText.getText().isEmpty() || objText.getText().isEmpty()) {
//					Alert alert = new Alert(Alert.AlertType.ERROR, "Please select map files!", ButtonType.OK);
//					alert.show();
//					return;
//				}
//				if (!(lsText.getText().endsWith(".gz") || lsText.getText().endsWith(".dat")) || !(objText.getText().endsWith(".gz") || objText.getText().endsWith(".dat"))) {
//					Alert alert = new Alert(Alert.AlertType.ERROR, "Map files need to be .gz or .dat format!", ButtonType.OK);
//					alert.show();
//					return;
//				}
//				try {
//					byte[] tileData = Files.readAllBytes(new File(lsText.getText()).toPath());
//					byte[] objData = Files.readAllBytes(new File(objText.getText()).toPath());
//
//					if (objText.getText().endsWith(".gz")) {
//						objData = GZIPCompressor.inflate317(objData);
//					}
//					if (lsText.getText().endsWith(".gz")) {
//						tileData = GZIPCompressor.inflate317(tileData);
//					}
//
//					xteaKeys.putAll(packMaps(regionX, regionY, tileData, objData, xteas));
//
//				} catch (Exception ex) {
//					ex.printStackTrace();
//				}
//			}
//
//			if(CacheLibrary.get().getIndex(OSRSIndices.MAPS).update(Selection.progressListener, xteaKeys)){
//				Alert alert = new Alert(Alert.AlertType.INFORMATION, "Maps packed successfully!", ButtonType.OK);
//				alert.show();
//            }
//
//		});
//		genXteas.setOnAction(event -> {
//			generateXTEAS();
//		});
//		canvas.widthProperty().bind(worldAnchor.widthProperty());
//		canvas.heightProperty().bind(worldAnchor.heightProperty());
//		regionIdSubmit.setOnAction((e) -> {
//			try {
//				int id = Integer.parseInt(regionIdField.getText());
//				renderer.setRegionId(id);
//			} catch (Exception ex) {
//				renderer.setRegionId(-1);
//			}
//		});
//	}
//
//	public int[] generateXTEAS() {
//		List<Integer> keys = Lists.newArrayList();
//		Random random = new Random();
//		keys.add(random.nextInt());
//		keys.add(random.nextInt());
//		keys.add(random.nextInt());
//		keys.add(random.nextInt());
//		xteaText.setText(keys.stream().map(String::valueOf).collect(Collectors.joining(",")));
//		return keys.stream().mapToInt(Integer::intValue).toArray();
//	}
//
//	public Map<Integer, int[]> packMaps(int regionX, int regionY, byte[] tileData, byte[] objData, int[] xteas) {
//		try {
//
//			String mapArchiveName = "m" + regionX + "_" + regionY;
//			String landArchiveName = "l" + regionX + "_" + regionY;
//
//			Archive mapArchive = CacheLibrary.get().getIndex(OSRSIndices.MAPS).getArchive(mapArchiveName);
//			boolean exists = Objects.nonNull(mapArchive);
//
//			if (exists) {
//				mapArchive.reset();
//			} else {
//				mapArchive = CacheLibrary.get().getIndex(OSRSIndices.MAPS).addArchive(mapArchiveName);
//			}
//			mapArchive.addFile(0, tileData);
//			mapArchive.flag();
//
//			Archive landArchive = CacheLibrary.get().getIndex(OSRSIndices.MAPS).getArchive(landArchiveName);
//			exists = Objects.nonNull(landArchive);
//
//			if (exists) {
//				landArchive.reset();
//			} else {
//				landArchive = CacheLibrary.get().getIndex(OSRSIndices.MAPS).addArchive(landArchiveName);
//			}
//			landArchive.addFile(0, objData);
//			landArchive.flag();
//
//			Map<Integer, int[]> xteaKeys = Maps.newHashMap();
//			xteaKeys.put(landArchive.getId(), xteas);
//
//			return xteaKeys;
//		} catch (Exception e) {
//			e.printStackTrace();
//			Alert alert = new Alert(Alert.AlertType.ERROR, "Failed to pack maps!", ButtonType.OK);
//			alert.show();
//			return Maps.newHashMap();
//		}
//	}
//
//	@Override
//	public void save() {
//
//	}
//
//	@Override
//	public ConfigEditorInfo getInfo() {
//		return null;
//	}
//}

File: SuitePlugins/MapPacker/src/main/kotlin/dev/dread/map/packer/MapPacker.kt
package dev.dread.map.packer

import dev.dread.map.crypto.XTEAProvider
import javafx.fxml.FXML
import javafx.scene.canvas.Canvas
import javafx.scene.control.*
import javafx.scene.control.Alert.AlertType
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import misc.CustomTab
import store.CacheLibrary
import store.cache.index.OSRSIndices
import store.plugin.extension.FXController
import store.utilities.GZIPCompressor
import suite.Main
import utility.ConfigEditorInfo
import utility.FilterMode
import utility.MultiMapEncoder
import utility.RetentionFileChooser
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths
import java.security.SecureRandom
import kotlin.math.floor

class MapPacker : FXController() {

    private val random = SecureRandom()

    private lateinit var tab: CustomTab
    private lateinit var xteaProvider: XTEAProvider

    override fun initialize(tab: CustomTab, refresh: Boolean, lastId: Int) {
        this.tab = tab
        regionType.selectedToggleProperty().addListener { observable, oldValue, newValue ->
            regionIdBox.isDisable = newValue != regionIdRadio;
            regionXYBox.isDisable = newValue != regionXYRadio;
            worldXYBox.isDisable = newValue != worldRadio;
        }
        loadType.selectedToggleProperty().addListener { observable, oldValue, newValue ->
            folderFileBox.isDisable = newValue != folderRadio;
            singleFileBox.isDisable = newValue == folderRadio;
        }

        saveButton()
        setupBrowseFolders()
    }

    private fun saveButton() {
        saveBtn.setOnAction { evt ->
            if (!::xteaProvider.isInitialized) {
                val file = RetentionFileChooser.showOpenDialog(
                    "Choose your xteas file",
                    Main.getSelection().stage.scene.window,
                    FilterMode.JSON
                )
                if (file != null && file.exists()) {
                    xteaProvider = XTEAProvider(file.toPath())
                } else {
                    Alert(AlertType.ERROR, "You must select a valid xteas file.", ButtonType.OK).show()
                    return@setOnAction
                }
            }

            val mapIndex = CacheLibrary.get().getIndex(OSRSIndices.MAPS)

            val regionX: Int
            val regionY: Int
            val regionId = regionID.text.toInt()
            if (regionType.selectedToggle === regionIdRadio) {
                regionX = regionId shr 8
                regionY = regionId and 0xFF
            } else if (regionType.selectedToggle === regionXYRadio) {
                regionX = regionXText.text.toInt()
                regionY = regionYText.text.toInt()
            } else {
                regionX = floor(worldXText.text.toInt() / 64.0).toInt()
                regionY = floor(worldYText.text.toInt() / 64.0).toInt()
            }

            // map of archive id, to xteas, we need these to update the cache
            val packingKeys = mutableMapOf<Int, IntArray>()

            // .pack file
            if (loadType.selectedToggle == folderRadio) {
                val text = mapFolderText.text

                // make sure the user has selected a file
                if (text.isEmpty()) {
                    Alert(AlertType.ERROR, "Please select a .pack file.", ButtonType.OK).show()
                    return@setOnAction
                }

                // make sure the file exists, and is a .pack file
                val file = File(text)
                if (!text.endsWith(".pack")) {
                    Alert(AlertType.ERROR, "You must select a .pack file", ButtonType.OK).show()
                    return@setOnAction
                }

                // make sure the file exists
                if (!file.exists()) {
                    Alert(AlertType.ERROR, "The file you selected does not exist.", ButtonType.OK).show()
                    return@setOnAction
                }

                try {
                    val chunks = MultiMapEncoder.decode(Files.readAllBytes(file.toPath()))
                    for (chunk in chunks) {
                        val newRegionX = (regionX + chunk.offsetX)
                        val newRegionY = (regionY + chunk.offsetY)
                        val regionId = newRegionX shl 8 or newRegionY
                        val newXtea = genXTEAKeys()
                        val archiveId = pack(newRegionX, newRegionY, chunk.tileMapData, chunk.objectMapData)
                        println("Packed region $regionId, archive id: $archiveId, xteas: ${newXtea.contentToString()}")
                        packingKeys[archiveId] = newXtea
                        xteaProvider.put(regionId, newXtea)
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                    Alert(AlertType.ERROR, "An error occurred while packing the map.", ButtonType.OK).show()
                    return@setOnAction
                }
            } else { // .dat/.gz file
                val lsText = lsText.text
                val objText = objText.text
                // make sure both files are selected
                if (lsText.isEmpty() || objText.isEmpty()) {
                    Alert(AlertType.ERROR, "Please select both the .ls and .obj files.", ButtonType.OK).show()
                    return@setOnAction
                }
                // make sure both files are .gz or .dat
                if (!(lsText.endsWith(".dat") || lsText.endsWith(".gz")) || !(objText.endsWith(".dat") || objText.endsWith(".gz"))) {
                    Alert(AlertType.ERROR, "The .ls file must be a .dat or .gz file.", ButtonType.OK).show()
                    return@setOnAction
                }
                try {
                    val tiles = Files.readAllBytes(Paths.get(lsText)).run(GZIPCompressor::inflate317)
                    val locs = Files.readAllBytes(Paths.get(objText)).run(GZIPCompressor::inflate317)
                    val xtea = genXTEAKeys()
                    val archiveId = pack(regionX, regionY, tiles, locs)
                    packingKeys[archiveId] = xtea
                    xteaProvider.put(regionId, xtea)
                } catch (e: Exception) {
                    e.printStackTrace()
                    Alert(AlertType.ERROR, "An error occurred while packing the map.", ButtonType.OK).show()
                    return@setOnAction
                }
            }

            if (mapIndex.update(packingKeys)) {
                // write the new xtea keys to the xteas file
                xteaProvider.save()
                Alert(AlertType.INFORMATION, "Map packed successfully.", ButtonType.OK).show()
            } else {
                Alert(AlertType.ERROR, "An error occurred while updating the cache.", ButtonType.OK).show()
            }
        }
    }

    private fun genXTEAKeys(): IntArray = random.ints(4).toArray()

    /**
     * Packs the map data into the cache.
     * @param regionX The region x coordinate.
     * @param regionY The region y coordinate.
     * @param tiles The tile data.
     * @param locs The location data.
     * @param xteas The xteas.
     * @return The archive id of the loc archive.
     */
    private fun pack(regionX: Int, regionY: Int, tiles: ByteArray, locs: ByteArray): Int {
        val index = CacheLibrary.get().getIndex(OSRSIndices.MAPS)

        val mapName = "m${regionX}_${regionY}"
        val mapArchive = index.addArchive(mapName, true)
        mapArchive.addFile(0, tiles)

        val locName = "l${regionX}_${regionY}"
        val locArchive = index.addArchive(locName, true)
        locArchive.addFile(0, locs)

        return locArchive.id
    }

    private fun setupBrowseFolders() {
        mapFolderBrowse.setOnAction { evt ->
            val file = RetentionFileChooser.showOpenDialog(
                Main.getSelection().stage.scene.window,
                FilterMode.PACK
            )
            if (file != null) {
                require(file.exists()) { "file doesn't exist! "}
                mapFolderText.text = file.absoluteFile.toString()
            }
        }
        lsBrowse.setOnAction { evt ->
            val file = RetentionFileChooser.showOpenDialog(
                Main.getSelection().stage.scene.window,
                FilterMode.DAT,
                FilterMode.GZIP
            )
            if (file != null) {
                lsText.text = file.absoluteFile.toString()
            }
        }
        objBrowse.setOnAction { evt ->
            val file = RetentionFileChooser.showOpenDialog(
                Main.getSelection().stage.scene.window,
                FilterMode.DAT,
                FilterMode.GZIP
            )
            if (file != null) {
                objText.text = file.absoluteFile.toString()
            }
        }
    }

    override fun save() = Unit

    override fun getInfo(): ConfigEditorInfo? = null

    @FXML
    private lateinit var folderRadio: RadioButton

    @FXML
    private lateinit var loadType: ToggleGroup

    @FXML
    private lateinit var folderFileBox: VBox

    @FXML
    private lateinit var mapFolderText: TextField

    @FXML
    private lateinit var mapFolderBrowse: Button

    @FXML
    private lateinit var singleFileRadio: RadioButton

    @FXML
    private lateinit var singleFileBox: VBox

    @FXML
    private lateinit var lsText: TextField

    @FXML
    private lateinit var lsBrowse: Button

    @FXML
    private lateinit var objText: TextField

    @FXML
    private lateinit var objBrowse: Button

    @FXML
    private lateinit var regionIdRadio: RadioButton

    @FXML
    private lateinit var regionType: ToggleGroup

    @FXML
    private lateinit var regionIdBox: HBox

    @FXML
    private lateinit var regionID: TextField

    @FXML
    private lateinit var regionXYRadio: RadioButton

    @FXML
    private lateinit var regionXYBox: HBox

    @FXML
    private lateinit var regionXText: TextField

    @FXML
    private lateinit var regionYText: TextField

    @FXML
    private lateinit var worldRadio: RadioButton

    @FXML
    private lateinit var worldXYBox: HBox

    @FXML
    private lateinit var worldXText: TextField

    @FXML
    private lateinit var worldYText: TextField

    @FXML
    private lateinit var xteaText: TextField

    @FXML
    private lateinit var genXteas: Button

    @FXML
    private lateinit var saveBtn: Button

    @FXML
    private lateinit var canvas: Canvas

    @FXML
    private lateinit var worldAnchor: AnchorPane

    @FXML
    private lateinit var regionIdField: TextField

    @FXML
    private lateinit var regionIdSubmit: Button

}